tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -527,6 +527,85 @@ can_sorted_merge_append(PlannerInfo *root, CompressionInfo *info, Chunk *chunk)
 	return merge_result;
 }
 
+/*
+ * This function adds per-chunk sorted paths for compressed chunks if beneficial. This has two
+ * advantages:
+ *
+ *  (1) Make ChunkAppend possible. If at least one chunk of a hypertable is uncompressed, PostgreSQL
+ * will generate a MergeAppend path in generate_orderedappend_paths() / create_merge_append_path()
+ * due to the existing pathkeys of the index on the uncompressed chunk. If all chunks are
+ * compressed, no path keys are present and no MergeAppend path is generated by PostgreSQL. In that
+ * case, the ChunkAppend optimization cannot be used because MergeAppend path can be promoted in
+ * ts_chunk_append_path_create(). Adding a sorted path with pathkeys makes ChunkAppend possible for
+ * these queries.
+ *
+ *  (2) Sorting on a per-chunk basis and merging / appending these results could be faster than
+ * sorting the whole input. Especially limit queries that use an ORDER BY that is compatible with
+ * the partitioning of the hypertable could be inefficiently executed otherwise. For example, an
+ * expensive query plan with a sort node on top of the append node could be chosen. Due to the sort
+ * node at the high level in the query plan and the missing ChunkAppend node (see (1)), all chunks
+ * are decompressed (instead of only the actually needed ones).
+ *
+ * The logic is inspired by PostgreSQL's add_paths_with_pathkeys_for_rel() function.
+ *
+ * Note: This function adds only non-partial paths. In parallel plans PostgreSQL prefers sorting
+ * directly under the gather (merge) node and the per-chunk sortings are not used in parallel plans.
+ * To save planning time, we therefore refrain from adding them.
+ */
+static void
+add_chunk_sorted_paths(PlannerInfo *root, RelOptInfo *chunk_rel, Hypertable *ht, Index ht_relid,
+					   Path *decompress_chunk_path, Path *child_path)
+{
+	if (root->parse->sortClause == NIL || root->sort_pathkeys == NIL ||
+		hypertable_is_distributed(ht))
+		return;
+
+	/* We are only interested in regular (i.e., non index) paths */
+	if (!IsA(child_path, Path))
+		return;
+
+	DecompressChunkPath *dcpath =
+		copy_decompress_chunk_path((DecompressChunkPath *) decompress_chunk_path);
+
+	/* Iterate over the sort_pathkeys and generate all possible useful sortings */
```

Comment:
Generating more paths here will amplify the planning issues with many chunks especially if they have distinct pathkeys as that will prevent them from getting purged by cheaper paths. Should the default path create an ordering according to orderby?

Reply:
We could add only one sorted path for all pathkeys. However, by adding all possible useful sortings (as PostgreSQL does in `add_paths_with_pathkeys_for_rel`) we allow to perform incremental sorting. However, even on the large `ht_chunk_compressed_2k` benchmark in tsbench, I see no regression with the current version of the PR.